mlir/include/mlir/Dialect/MIOpen/XdlopsCodeSelection.h

```diff
@@ -467,221 +440,86 @@ struct XdlopsCodeSelection {
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 32;
-      num_output_blks = 1;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 32;
-      n = 32;
-      k = 4;
-      cycles = 64;
-      k_base = 2;
-    } else if (instr == amdgpu::MFMAInstr::f32_16x16x8bf16) {
-      group_size = 4;
-      num_groups_blk = 1;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 16;
-      num_output_blks = 1;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 16;
-      n = 16;
-      k = 8;
-      cycles = 32;
-      k_base = 2;
-    } else if (instr == amdgpu::MFMAInstr::f32_32x32x2f32) {
-      group_size = 4;
-      num_groups_blk = 1;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 16;
-      num_output_blks = 4;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 16;
-      n = 16;
-      k = 2;
-      cycles = 32;
-      k_base = 2;
-    } else if (instr == amdgpu::MFMAInstr::f32_4x4x2bf16) {
-      group_size = 4;
-      num_groups_blk = 1;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 64;
-      num_output_blks = 1;
-      num_regs_xdlops = 4;
-      m = 4;
-      n = 64;
-      k = 2;
-      cycles = 8;
-      k_base = 2;
-    } else if (instr == amdgpu::MFMAInstr::i32_32x32x8i8) {
-      group_size = 4;
-      num_groups_blk = 4;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 32;
-      num_output_blks = 1;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 32;
-      n = 32;
-      k = 8;
-      cycles = 64;
-      k_base = 4;
-    } else if (instr == amdgpu::MFMAInstr::i32_16x16x16i8) {
-      group_size = 4;
-      num_groups_blk = 1;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 16;
-      num_output_blks = 1;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 16;
-      n = 16;
-      k = 16;
-      cycles = 32;
-      k_base = 4;
-    } else {
-      llvm_unreachable("Unsupported case as mfmaInstr not selected!\n");
-    }
+    // Derived properties of the individual MFMA. These are computed here
+    // and used in places throughout the code and may not all be needed.
+    int64_t kPerInputArg =
+        math_util::integer_divide_ceil(waveSize, mfmaNonKDim * blocksMfma);
+    // k_base is the number of times you need to step in the k dimension on each
+    // lane in a wave.
+    int64_t k_base = k / kPerInputArg;
+
+    // Number of logical values each thread needs to pass in to the MFMA in
+    // order for the correct number of input values to be passed to the MFMA.
+    int64_t nInputsToMfma = (mfmaNonKDim * blocksMfma * k) / waveSize;
+    Type argType = nInputsToMfma == 1
+                       ? dataType
+                       : VectorType::get({nInputsToMfma}, dataType);
+
+    int64_t nOutputsOfMfma =
+        (mfmaNonKDim * mfmaNonKDim * blocksMfma) / waveSize;
```

Comment:
Reduction occupies fewer registers than the corresponding non-reduction, yeah

Reply:
That makes sense, thanks! So `nOutputsOfMfma` has the special condition to not be completely filled up towards lane 63, but we occupy the same amount of registers regardless. Since you are describing the number of registers occupied, it doesn't matter.

Reply:
There is filling up to lane 63. The computation is this: We know each MFMA will produce mfmaNonKDim (aka m_mfma) * mfmaNonKDim (aka n_mfma) * blocksMfma outputs, and that those outputs will be put into some number of outputs, with each output containing waveSize (aka 64) values.

Reply:
Gotcha now, thank you!